### PR TITLE
Add @no_dry_run to Node.power_cycle

### LINKED ISSUE #404: Node_power_cycle produces error (dry_run = true)
Calling "node_power_cycle" produces the following error message and traceback.
I have dry_run = true in haas.cfg

(.venv)kamfonik@dandini-ubuntu:~/Desktop/haas$ haas node_power_cycle 1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
DEBUG:moc.rest:Recieved api call node_power_cycle(u'1')
INFO:werkzeug:127.0.0.1 - - [24/Mar/2015 09:00:54] "POST /node/1/power_cycle HTTP/1.1" 500 -
DEBUG:requests.packages.urllib3.connectionpool:"POST /node/1/power_cycle HTTP/1.1" 500 291
Unexpected status code: 500
Response text:
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>500 Internal Server Error</title>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error and was unable to complete your request.  Either the server is overloaded or there is an error in the application.</p>

ERROR:werkzeug:Error on request:
Traceback (most recent call last):
  File "/home/kamfonik/Desktop/haas/.venv/lib/python2.7/site-packages/werkzeug/serving.py", line 177, in run_wsgi
    execute(self.server.app)
  File "/home/kamfonik/Desktop/haas/.venv/lib/python2.7/site-packages/werkzeug/serving.py", line 165, in execute
    application_iter = app(environ, start_response)
  File "/home/kamfonik/Desktop/haas/.venv/lib/python2.7/site-packages/werkzeug/local.py", line 224, in application
    return ClosingIterator(app(environ, start_response), self.cleanup)
  File "/home/kamfonik/Desktop/haas/moc/rest.py", line 201, in wsgi_handler
    response = request_handler(Request(environ))
  File "/home/kamfonik/Desktop/haas/moc/rest.py", line 182, in request_handler
    response_body = f(*positional_args)
  File "/home/kamfonik/Desktop/haas/haas/api.py", line 237, in node_power_cycle
    if not node.power_cycle():
  File "/home/kamfonik/Desktop/haas/haas/model.py", line 168, in power_cycle
    self._ipmitool(['chassis', 'bootdev', 'pxe'])
  File "/home/kamfonik/Desktop/haas/haas/model.py", line 157, in _ipmitool
    '-H', self.ipmi_host] + args)
  File "/usr/lib/python2.7/subprocess.py", line 522, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1327, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory

### PR DESCRIPTION
**Pull request by zenhack**
Fixes #404 

@kamfonik, can you check if this fixes the problem?

### CODE REVIEW

**Comment by kamfonik** [comment]
I think so? See the output below - is it supposed to throw the ServerError when no_dry_run returns "None" ? 

command terminal:
(.venv)kamfonik@dandini-ubuntu:~/Desktop/haas$ haas node_power_cycle 0
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
DEBUG:requests.packages.urllib3.connectionpool:"POST /node/0/power_cycle HTTP/1.1" 500 62
Unexpected status code: 500
Response text:
{"msg": "Could not power cycle node 0", "type": "ServerError"}

server terminal:
DEBUG:moc.rest:Recieved api call node_power_cycle(u'0')
INFO:haas.dev_support:dry run, not executing: haas.model.power_cycle(*(Node<u'0'>,),**{})
DEBUG:moc.rest:Invalid call to api function node_power_cycle, raised exception: ServerError(u'Could not power cycle node 0',)
INFO:werkzeug:127.0.0.1 - - [24/Mar/2015 16:49:29] "POST /node/0/power_cycle HTTP/1.1" 500 -

**Comment by zenhack** [comment]
That *might* be the right behavior? `<sacrasm>`Good thing we have a detailed specification to tell us!`</sarcasm>` @gsilvis what do you think of this?

Everything is executing as it should (or should not), and HaaS is (arguably correctly) reporting that it's unable to power cycle the node -- the `None` return value isn't totally rediculous in this case, but there's a question as to how the success/failure reporting should work when the thing is disabled.

**Comment by gsilvis** [comment]
This behavior is definitely wrong, I'd say.  We need to make it so that the ``no_dry_run``'d version of the function reports something that looks like success, or at least something that doesn't look like failure.

**Comment by zenhack** [comment]
> This behavior is definitely wrong, I'd say. We need to make it so that
> the no_dry_run'd version of the function reports something that looks
> like success, or at least something that doesn't look like failure.

Yeah, I think so too. What if `power_cycle` raised the exception itself?  
This way `dry_run` would suppress that as well.

**Comment by gsilvis** [comment]
I like exceptions---that plan seems good to me.

**Comment by zenhack** [comment]
@gsilvis, as we discussed, moving the raise into `haas.model` was going to cause a circular dependency, so I took the opportunity to reorganize things a bit. There are a few style/doc fixes in here as well -- I noticed things by virtue of looking at nearby code. I've tried to made the individual commits coherent, which should aid in understanding patch.

**Comment by zenhack** [comment]
@lokI8, @henn, can one of you review this as well?

File: haas/model.py [hunk 1]
```diff
@@ -172,7 +174,9 @@ def power_cycle(self):
             # just turn it on in that case. This way we can save power by
             # turning things off without breaking the HaaS.
             status = self._ipmitool(['chassis', 'power', 'on'])
-        return status == 0
+            if status != 0:
+                # If it still doesn't work, then it's a real error:
+                raise OBMError('Could not power cycle node %s' % node.label)
```
Comment (gsilvis) [inline haas/model.py]: Suggest::

    if 0 != self._ipmitool(['chassis', 'power', 'cycle']):
        return
    if 0 != self._ipmitool(['chassis', 'power', 'on']):
        return
    raise OBMError('Could not power cycle node %s' % node.label)

**Comment by gsilvis** [comment]
This looks pretty good to me.  +1

**Comment by zenhack** [comment]
@henn, @lokI8, can one of you do a second review?

**Comment by lokI8** [comment]
+1, merging

**Pull request merged**